src/preact/base-element.js

```diff
@@ -131,16 +131,124 @@ export class PreactBaseElement extends BaseElement {
     return !this.usesLoading();
   }
 
-  /** @param {!Element} element */
+  /**
+   * Override to provide the Component definition.
+   *
+   * @type {import('preact').FunctionComponent}
+   * @return {ReturnType<import('preact').FunctionComponent>}
+   */
+  static Component() {
+    devAssert(false, 'Must provide Component');
+  }
+
+  /**
+   * If default props are static, this can be used instead of init().
+   * @type {!JsonObject|undefined}
+   */
+  static staticProps = undefined;
+
+  /**
+   * @type {*}
+   */
+  static useContexts = mode.isLocalDev() ? Object.freeze([]) : [];
+
+  /**
+   * Whether the component implements a loading protocol.
+   *
+   * @type {boolean}
+   */
+  static loadable = false;
+
+  /**
+   * Whether a component should be unloaded for `pauseCallback`.
+   *
+   * @type {boolean}
+   */
+  static unloadOnPause = false;
+
+  /**
+   * An override to specify that the component requires `layoutSizeDefined`.
+   * This typically means that the element's `isLayoutSupported()` is
+   * implemented via `isLayoutSizeDefined()`, and this is how the default
+   * `isLayoutSupported()` is implemented when this flag is set.
+   *
+   * @type {boolean}
+   */
+  static layoutSizeDefined = false;
+
+  /**
+   * The tag name, e.g. "div", "span", time" that should be used as a replacement
+   * node for Preact rendering. This is the node that Preact will diff with
+   * with specified, instead of rendering a new node. Only applicable to light-DOM
+   * mapping styles.
+   *
+   * @type {string}
+   */
+  static lightDomTag = '';
+
+  /**
+   * Whether this element uses "templates" system.
+   *
+   * @type {boolean}
+   */
+  static usesTemplate = false;
+
+  /**
+   * The CSS for shadow stylesheets.
+   *
+   * @type {?string}
+   */
+  static shadowCss = null;
+
+  /**
+   * Whether this element uses Shadow DOM.
+   *
+   * @type {boolean}
+   */
+  static usesShadowDom = false;
+
+  /**
+   * Enabling detached mode alters the children to be rendered in an
+   * unappended container. By default the children will be attached to the DOM.
+   *
+   * @type {boolean}
+   */
+  static detached = false;
+
+  /**
+   * This enables the 'delegatesFocus' option when creating the shadow DOM for
+   * this component.  A key feature of 'delegatesFocus' set to true is that
+   * when elements within the shadow DOM gain focus, the focus is also applied
+   * to the host element.
+   */
+  static delegatesFocus = false;
+
+  /**
+   * Provides a mapping of Preact prop to AmpElement DOM attributes.
+   *
+   * @type {Object<string, AmpElementProp>}
+   */
+  static props = {};
+
+  /** @param {AmpElement} element */
   constructor(element) {
     super(element);
 
     /** @private {!JsonObject} */
     this.defaultProps_ = dict({
       'loading': Loading_Enum.AUTO,
+
+      /**
+       * @param {import('#core/constants/ready-state').ReadyState_Enum} state
+       * @param {*} opt_failure
```

Comment:
This is treated as an Error down the call stack. Maybe should be typed as one (and definitely should be typed as optional, which it is elsewhere in this file

Reply:
Done!